pydantic/networks.py

```diff
@@ -86,136 +88,294 @@ def __hash__(self) -> int:
-* Host required
-"""
-HttpUrl = Annotated[Url, UrlConstraints(max_length=2083, allowed_schemes=['http', 'https'])]
-"""A type that will accept any http or https URL.
-
-* TLD not required
-* Host required
-* Max length 2083
-
-```py
-from pydantic import BaseModel, HttpUrl, ValidationError
-
-class MyModel(BaseModel):
-    url: HttpUrl
-
-m = MyModel(url='http://www.example.com')  # (1)!
-print(m.url)
-#> http://www.example.com/
-
-try:
-    MyModel(url='ftp://invalid.url')
-except ValidationError as e:
-    print(e)
-    '''
-    1 validation error for MyModel
-    url
-      URL scheme should be 'http' or 'https' [type=url_scheme, input_value='ftp://invalid.url', input_type=str]
-    '''
-
-try:
-    MyModel(url='not a url')
-except ValidationError as e:
-    print(e)
-    '''
-    1 validation error for MyModel
-    url
-      Input should be a valid URL, relative URL without a base [type=url_parsing, input_value='not a url', input_type=str]
-    '''
-```
-
-1. Note: mypy would prefer `m = MyModel(url=HttpUrl('http://www.example.com'))`, but Pydantic will convert the string to an HttpUrl instance anyway.
-
-"International domains" (e.g. a URL where the host or TLD includes non-ascii characters) will be encoded via
-[punycode](https://en.wikipedia.org/wiki/Punycode) (see
-[this article](https://www.xudongz.com/blog/2017/idn-phishing/) for a good description of why this is important):
-
-```py
-from pydantic import BaseModel, HttpUrl
-
-class MyModel(BaseModel):
-    url: HttpUrl
-
-m1 = MyModel(url='http://puny£code.com')
-print(m1.url)
-#> http://xn--punycode-eja.com/
-m2 = MyModel(url='https://www.аррӏе.com/')
-print(m2.url)
-#> https://www.xn--80ak6aa92e.com/
-m3 = MyModel(url='https://www.example.珠宝/')
-print(m3.url)
-#> https://www.example.xn--pbt977c/
-```
-
-
-!!! warning "Underscores in Hostnames"
-    In Pydantic, underscores are allowed in all parts of a domain except the TLD.
-    Technically this might be wrong - in theory the hostname cannot have underscores, but subdomains can.
-
-    To explain this; consider the following two cases:
-
-    - `exam_ple.co.uk`: the hostname is `exam_ple`, which should not be allowed since it contains an underscore.
-    - `foo_bar.example.com` the hostname is `example`, which should be allowed since the underscore is in the subdomain.
-
-    Without having an exhaustive list of TLDs, it would be impossible to differentiate between these two. Therefore
-    underscores are allowed, but you can always do further validation in a validator if desired.
-
-    Also, Chrome, Firefox, and Safari all currently accept `http://exam_ple.com` as a URL, so we're in good
-    (or at least big) company.
-"""
-AnyWebsocketUrl = Annotated[Url, UrlConstraints(allowed_schemes=['ws', 'wss'])]
-"""A type that will accept any ws or wss URL.
+    @classmethod
+    def __get_pydantic_core_schema__(cls, source: type[Any], handler: GetCoreSchemaHandler) -> core_schema.CoreSchema:
+        if cls is source:
```

Comment:
This means that `_BaseUrl` would be used as an annotation?

```python
class Model(BaseModel):
    a: _BaseUrl
```

Perhaps you meant using `issubclass`?